src/ql/ir/compat/kernel.cc

```diff
@@ -1055,6 +1055,22 @@ void Kernel::gate(
     gates.extend(u.get_decomposition(qubits));
 }
 
+// adding state prepration / (arbitrary) qubit initialisation to kernel
+void Kernel::state_prep(
+    const Vec<Complex> &array,
+    const Vec<UInt> &qubits
+) {
+    QL_DOUT("Preparing state with array " << array);
+    cycles_valid = false;
```

Comment:
I was planning to make it do auto-normalization, because doing that manually becomes annoying quickly, but I dont remember if I actually implemented it. If I didnt, then there should be a check for it indeed!

Reply:
p.s. I once understood that QX emulator does automatic re-normalization of state vectors. This was introduced after it was observed that repeated Hadamard gates did not bring the state vector back to its original state because of the limited precision of sqrt(2) in the operation. Nader then introduced renormalization. I do not were it is in the QX code base but could be used an an inspiration perhaps?

Reply:
Very interesting, thanks for the context